src/modules/lua.c

```diff
@@ -191,19 +212,88 @@ mtev_lua_lmc_free(lua_module_closure_t *lmc) {
                      (const char*)&lmc->lua_state, sizeof(lmc->lua_state),
                      free, NULL);
     pthread_mutex_unlock(&mtev_lua_states_lock);
+    if(lmc->timer) {
+      if (timer_delete(lmc->_timer) == -1) {
+        mtevL(mtev_error, "timer_delete failed: %s", strerror(errno));
+      }
+    }
   }
   free(lmc);
 }
 
 static __thread lua_State *tls_active_lua_state = NULL;
 static __thread bool assist_fired;
 
+static void
+mtev_lua_timer_start(timer_t *lua_timer, struct timeval *diff) {
+  if(lua_timer == NULL) return;
+
+  mtevL(nldeb, "Starting lua timer: %p for %ld.%06ld\n", lua_timer, diff->tv_sec, diff->tv_usec);
+  struct itimerspec ispec;
+  memset(&ispec, 0, sizeof(ispec));
+  ispec.it_value.tv_sec = diff->tv_sec;
+  ispec.it_value.tv_nsec = diff->tv_usec * 1000;
+  if(timer_settime(*lua_timer, 0, &ispec, NULL) != 0)
+    mtevL(mtev_error, "timer_settime failed: %s\n", strerror(errno));
+}
+
+static void
+mtev_lua_timer_stop(timer_t *lua_timer) {
+  if(lua_timer == NULL) return;
+
+  mtevL(nldeb, "Stopping lua timer: %p\n", lua_timer);
+  struct itimerspec ispec;
+  memset(&ispec, 0, sizeof(ispec));
+  if(timer_settime(*lua_timer, 0, &ispec, NULL) != 0)
+    mtevL(mtev_error, "timer_settime failed: %s\n", strerror(errno));
+}
+static void lstop (lua_State *L, lua_Debug *ar) {
+  (void)ar;  /* unused arg. */
+  lua_sethook(L, NULL, 0, 0);  /* reset hook */
+  mtev_stacktrace(mtev_error);
+  luaL_error(L, "interrupted!");
+}
+
+static void
+mtev_lua_timer_fire(int sig, siginfo_t *info, void *c) {
+  (void)sig;
+  (void)info;
+  (void)c;
+  mtevL(nldeb, "timer fired: lua %p\n", tls_active_lua_state);
+
+  if(tls_active_lua_state) {
+    lua_sethook(tls_active_lua_state, lstop, LUA_MASKCALL | LUA_MASKRET | LUA_MASKCOUNT,1);
+  }
+}
+
 int
-mtev_lua_resume(lua_State *L, int a) {
+mtev_lua_resume(lua_State *L, int a, mtev_lua_resume_info_t *ri) {
   int rv;
+  struct timeval diff = { .tv_sec = 0, .tv_usec = 0 };
   lua_State *previous = tls_active_lua_state;
   tls_active_lua_state = L;
+  if(eventer_heartbeat_deadline(NULL, &diff)) {
+    sub_timeval(diff, (struct timeval){ .tv_sec = 2, .tv_usec = 0 }, &diff);
+    if(diff.tv_sec < 0 || diff.tv_usec < 0) {
+      mtevL(nlerr, "lua attempting to resume with no time left before watchdog\n");
+      mtev_stacktrace(nlerr);
+      diff.tv_sec = 0;
+      diff.tv_usec = 1;
+    }
+  }
+  if(ri && ri->lmc && !(ri->lmc->interrupt_time.tv_sec == 0 && ri->lmc->interrupt_time.tv_usec == 0)) {
+    /* if we have an interrupt time specified for this lmc, then we should potentially reduce to it */
+    mtevL(mtev_error, "interrupt_time set...\n");
```

Comment:
Should this perhaps be a debug log? I can't see what value it has in the normal error log.